src/messages/mod.rs

```diff
@@ -298,17 +311,59 @@ impl SignedRoutingMessage {
         }
     }
 
-    /// Adds all signatures from the given message, without validating them.
-    pub fn add_signature_shares(&mut self, mut msg: Self) {
-        if self.content.src.is_multiple() {
-            if let (
-                SecurityMetadata::Partial(self_partial),
-                SecurityMetadata::Partial(other_partial),
-            ) = (&mut self.security_metadata, &mut msg.security_metadata)
-            {
-                self_partial.shares.append(&mut other_partial.shares);
+    /// Merge the signature shares of the two messages and take the longer of the proofs.
+    pub fn merge_security_metadata(
+        &mut self,
+        other: Self,
+    ) -> Result<(), MergeSecurityMetadataError> {
+        use MergeSecurityMetadataErrorKind::*;
+
+        if self.content != other.content {
+            return Err(MergeSecurityMetadataError {
+                kind: ContentMismatch,
+                other_message: other,
+            });
+        }
+
+        let self_md = match &mut self.security_metadata {
+            SecurityMetadata::Partial(md) => md,
+            _ => {
+                return Err(MergeSecurityMetadataError {
+                    kind: NotPartial,
+                    other_message: other,
+                })
             }
+        };
+
+        let mut other_md = match other.security_metadata {
+            SecurityMetadata::Partial(md) => md,
+            _ => {
+                return Err(MergeSecurityMetadataError {
+                    kind: NotPartial,
+                    other_message: other,
+                })
+            }
+        };
+
+        if self_md.pk_set != other_md.pk_set {
+            return Err(MergeSecurityMetadataError {
+                kind: PublicKeySetMismatch,
+                other_message: SignedRoutingMessage::from_parts(
+                    other.content,
+                    SecurityMetadata::Partial(other_md),
+                ),
+            });
+        }
+
+        // TODO: should we also check the proof chains are compatible?
```

Comment:
Considering that TODO, would it make sense to postpone the commit `Take the longest proof chain when accumulating message signatures` to a later PR that would address that too? Currently because there are no malicious node any chain will be trusted.

Reply:
Yes. Also I'm no longer sure just taking the longer proof is sufficient. Say we have two messages, both have proof of length 1, but one is at  version 7 and the other 8. The current code will prefer the one that is already in the accumulator, but I think we should take the lower of the two (not 100% sure about this though).

Reply:
Did you mean to remove that commit?